homeassistant/components/samsungtv/config_flow.py

```diff
@@ -96,81 +138,84 @@ async def async_step_import(self, user_input=None):
     async def async_step_user(self, user_input=None):
         """Handle a flow initialized by the user."""
         if user_input is not None:
-            ip_address = await self.hass.async_add_executor_job(
-                _get_ip, user_input[CONF_HOST]
-            )
-
-            await self.async_set_unique_id(ip_address)
-            self._abort_if_unique_id_configured()
-
-            self._host = user_input.get(CONF_HOST)
-            self._ip = self.context[CONF_IP_ADDRESS] = ip_address
-            self._name = user_input.get(CONF_NAME)
+            self._host = user_input[CONF_HOST]
+            self._name = user_input[CONF_NAME]
             self._title = self._name
 
-            result = await self.hass.async_add_executor_job(self._try_connect)
+            self._abort_if_already_configured()
+
+            await self.hass.async_add_executor_job(self._try_connect)
 
-            if result != RESULT_SUCCESS:
-                return self.async_abort(reason=result)
             return self._get_entry()
 
         return self.async_show_form(step_id="user", data_schema=DATA_SCHEMA)
 
-    async def async_step_ssdp(self, discovery_info):
-        """Handle a flow initialized by discovery."""
-        host = urlparse(discovery_info[ATTR_SSDP_LOCATION]).hostname
-        ip_address = await self.hass.async_add_executor_job(_get_ip, host)
-
-        self._host = host
-        self._ip = self.context[CONF_IP_ADDRESS] = ip_address
-        self._manufacturer = discovery_info.get(ATTR_UPNP_MANUFACTURER)
-        self._model = discovery_info.get(ATTR_UPNP_MODEL_NAME)
-        self._name = f"Samsung {self._model}"
-        self._id = discovery_info.get(ATTR_UPNP_UDN)
-        self._title = self._model
+    async def async_step_ssdp(self, user_input=None):
+        """Handle a flow initialized by ssdp discovery."""
+        self._host = urlparse(user_input[ATTR_SSDP_LOCATION]).hostname
+        self._id = user_input.get(ATTR_UPNP_UDN)
 
         # probably access denied
         if self._id is None:
             return self.async_abort(reason=RESULT_AUTH_MISSING)
         if self._id.startswith("uuid:"):
             self._id = self._id[5:]
 
-        await self.async_set_unique_id(ip_address)
-        self._abort_if_unique_id_configured(
-            {
-                CONF_ID: self._id,
-                CONF_MANUFACTURER: self._manufacturer,
-                CONF_MODEL: self._model,
-            }
-        )
+        await self.async_set_unique_id(self._id)
+        await self.hass.async_add_executor_job(self._get_and_check_device_info)
+
+        self._manufacturer = user_input.get(ATTR_UPNP_MANUFACTURER)
+        if not self._model:
+            self._model = user_input.get(ATTR_UPNP_MODEL_NAME)
+        self._name = f"{self._manufacturer} {self._model}"
+        self._title = self._model
+
+        await self.hass.async_add_executor_job(self._abort_if_already_configured())
```

Comment:
I recommend not running this whole method in sync context, just run the host lookup in sync context.